Src/Support/Google.Apis.Auth/OAuth2/AwsExternalAccountCredential.AwsSecurityCredential.cs

```diff
@@ -0,0 +1,144 @@
+            /// Returns null if the credential URL is null or empty.
+            /// </summary>
+            internal static async Task<AwsSecurityCredentials> MaybeFromMetadataAsync(AwsMetadataServerClient metadataClient, string credentialUrl)
+            {
+                // https://docs.aws.amazon.com/AWSEC2/latest/UserGuide/iam-roles-for-amazon-ec2.html#instance-metadata-security-credentials
+                if (!string.IsNullOrEmpty(credentialUrl)
+                    // Build the actual credential endpoint.
+                    && await BuildSecurityCredentialsEndpointAsync().ConfigureAwait(false) is string credentialEndpoint
+                    // Fetch the credential JSON
+                    && await metadataClient.FetchMetadataAsync(credentialEndpoint).ConfigureAwait(false) is string credentialJson
+                    && credentialJson != "")
+                {
+                    // Deserialize the credentials
+                    var deserializedCredentials = NewtonsoftJsonSerializer.Instance.Deserialize<AwsSecurityCredentialsResponse>(credentialJson);
+
+                    if (deserializedCredentials.IsSuccess
+                        && !string.IsNullOrEmpty(deserializedCredentials.AccessKeyId)
+                        && !string.IsNullOrEmpty(deserializedCredentials.SecretAccessKey))
+                    {
+                        return new AwsSecurityCredentials(
+                            deserializedCredentials.AccessKeyId, deserializedCredentials.SecretAccessKey,
+                            deserializedCredentials.Token == "" ? null : deserializedCredentials.Token);
+                    }
+
+                    throw new InvalidOperationException(
+                        $"The request to the metadata server security credentials endpoint {credentialEndpoint} was not succesful." +
+                        $"the response was {credentialJson}");
+                }
+
+                return null;
+
+                async Task<string> BuildSecurityCredentialsEndpointAsync()
+                {
+                    // https://github.com/aws/aws-sdk-java/blob/ccfd63c097874e3a1e9ffda7bf171769b6bd3b21/aws-java-sdk-core/src/main/java/com/amazonaws/auth/InstanceMetadataServiceCredentialsFetcher.java#L55
+                    // Fetch the list of roles attached to the instance from the credential URL as given.
+                    if (await metadataClient.FetchMetadataAsync(credentialUrl).ConfigureAwait(false) is string metadataRoleList)
+                    {
+                        // Let's use StringReader so that we don't have to worry about which end of line charater(s) to use.
+                        using var reader = new StringReader(metadataRoleList);
+                        if (await reader.ReadLineAsync().ConfigureAwait(false) is string role && role != "")
+                        {
+                            // Attach the role to the given credential URL to fetch the actual credentials.
```

Comment:
The credentialUrl is the one in the config file, yes? So it's "somewhat reasonable" for it to sometimes end with / and sometimes not? (Just wondering if we really need this.)